File: samples/replicatedentity-examples/src/main/java/com/example/replicated/set/domain/SomeSet.java
package com.example.replicated.set.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntityContext;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedSet;
import com.example.replicated.set.SomeSetApi;
import com.google.protobuf.Empty;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class SomeSet extends AbstractSomeSet {
  @SuppressWarnings("unused")
  private final String entityId;

  public SomeSet(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> add(
      ReplicatedSet<SomeSetDomain.SomeElement> set, SomeSetApi.AddElement command) {
    SomeSetDomain.SomeElement element = // <1>
        SomeSetDomain.SomeElement.newBuilder().setValue(command.getElement().getValue()).build();
    set.add(element); // <2>
    return effects()
        .update(set) // <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> remove(
      ReplicatedSet<SomeSetDomain.SomeElement> set, SomeSetApi.RemoveElement command) {
    SomeSetDomain.SomeElement element = // <1>
        SomeSetDomain.SomeElement.newBuilder().setValue(command.getElement().getValue()).build();
    set.remove(element); // <2>
    return effects()
        .update(set) // <3>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeSetApi.CurrentElements> get(
      ReplicatedSet<SomeSetDomain.SomeElement> set, SomeSetApi.GetElements command) {
    List<SomeSetApi.Element> elements =
        set.stream() // <1>
            .map(element -> SomeSetApi.Element.newBuilder().setValue(element.getValue()).build())
            .sorted(Comparator.comparing(SomeSetApi.Element::getValue))
            .collect(Collectors.toList());

    SomeSetApi.CurrentElements currentElements =
        SomeSetApi.CurrentElements.newBuilder().addAllElements(elements).build();

    return effects().reply(currentElements);
  }
  // end::get[]
}

File: samples/scala-replicated-examples/src/main/scala/com/example/replicated/set/domain/SomeSet.scala
package com.example.replicated.set.domain

import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedSet
import com.example.replicated.set
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** A replicated entity. */
class SomeSet(context: ReplicatedEntityContext) extends AbstractSomeSet {


  /** Command handler for "Add". */
  def add(currentData: ReplicatedSet[String], addElement: set.AddElement): ReplicatedEntity.Effect[Empty] =
    effects.error("The command handler for `Add` is not implemented, yet")

  /** Command handler for "Remove". */
  def remove(currentData: ReplicatedSet[String], removeElement: set.RemoveElement): ReplicatedEntity.Effect[Empty] =
    effects.error("The command handler for `Remove` is not implemented, yet")

  /** Command handler for "Get". */
  def get(currentData: ReplicatedSet[String], getElements: set.GetElements): ReplicatedEntity.Effect[set.CurrentElements] =
    effects.error("The command handler for `Get` is not implemented, yet")

}
